designer/server/src/main/scala/pl/touk/nussknacker/ui/api/ManagementResources.scala

```diff
@@ -245,28 +246,23 @@ class ManagementResources(val processAuthorizer: AuthorizeProcess,
             (post & processId(processName) & entity(as[CustomActionRequest])) { (process, req) =>
               val params = req.params.getOrElse(Map.empty)
               complete {
-                customActionInvokerService.invokeCustomAction(req.actionName, process, params)
-                  .flatMap {
-                    case res@Right(_) =>
-                      toHttpResponse(CustomActionResponse(res))(StatusCodes.OK)
-                    case res@Left(err) =>
-                      val response = toHttpResponse(CustomActionResponse(res)) _
-                      err match {
-                        case _: CustomActionFailure => response(StatusCodes.InternalServerError)
-                        case _: CustomActionInvalidStatus => response(StatusCodes.Forbidden)
-                        case _: CustomActionForbidden => response(StatusCodes.Forbidden)
-                        case _: CustomActionNotImplemented => response(StatusCodes.NotImplemented)
-                        case _: CustomActionNonExisting => response(StatusCodes.NotFound)
-                      }
-                  }
+                customActionInvokerService
+                  .invokeCustomAction(req.actionName, process, params)
+                  .map(_ => HttpResponse(status = StatusCodes.OK))
+                  .recover(customActionErrorToHttp orElse EspErrorToHttp.errorToHttp)
               }
             }
           }
       }
   }
 
-  private def toHttpResponse[A: Encoder](a: A)(code: StatusCode): Future[HttpResponse] =
-    Marshal(a).to[MessageEntity].map(en => HttpResponse(entity = en, status = code))
+  private def customActionErrorToHttp : PartialFunction[Throwable, HttpResponse] = {
```

Comment:
WDYT about making `EspErrorToHttp.errorToHttp` responsible for that? The only thing that we have to do is to make specific errors extending one of the `EspError`

Reply:
Good point. Fixed.